### PR TITLE
[InstCombine] Add log-pow simplification for FP exponent edge case.

### LINKED ISSUE #76549: clang is suboptimal for `log(pow(x, y))` with fast-math
Test case:
https://godbolt.org/z/1fvYPxY37
```c
#include<math.h>

double test1(double x) { 
  return log(pow(x, -3)); 
}

double test2(double x) { 
  return -3 * log(x); 
}
```

My expectation is that the output of `test1` will be the same as `test2`.

### PR DESCRIPTION
**Pull request by tanmaysachan**
Fixes https://github.com/llvm/llvm-project/issues/76549

The cause of the optimization miss was -
1. `optimizePow` converting almost integer FP exponents to integer, and turning `pow` to `powi`.
2. `optimizeLog` not accepting `Intrinsic::powi` as a target.

This patch converts constantInt back to constantFP where applicable and adds a test.

### CODE REVIEW

File: llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:2507]: If the `exponent` is not a constant, you will create an illegal fmul with an integer LHS.
Please add the following negative test and pre-commit these tests in a separate commit.
```
define double @log_pow_nonconst(double %x, i32 %y) {
  %pow = call fast double @llvm.powi.f64.i32(double %x, i32 %y)
  %log = call fast double @log(double %pow)
  ret double %log
}
```

Another question: why don't we fold `log(powi(x,y)) -> (sitofp y) * log(x)` even if `Y` is not a constant? I think it is also profitable if `powi` is only used by the `log`.

File: llvm/test/Transforms/InstCombine/log-pow.ll
Comment (dtcxzyw) [inline llvm/test/Transforms/InstCombine/log-pow.ll:21]:
```suggestion
  %pow = call fast double @llvm.powi.f64.i32(double %x, i32 -3)
```

File: llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:2507]:
```suggestion
      Y = ConstantFoldCastOperand(Instruction::SIToFP, C, Log->getType(), DL);
      if (!Y) return nullptr;
```
It would be better than `C->getSExtValue()` since it can handle integers wider than 64 bits.

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp
Comment (tanmaysachan) [reply llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:2507]: @dtcxzyw I did a standard sitofp initially, but that adds a casting instruction right? In the compiled asm there's an addition `scvtf` if using a non-constant cast.

I'll fix the illegal fmul, sorry about that.

Comment (dtcxzyw) [reply llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:2507]: > @dtcxzyw I did a standard sitofp initially, but that adds a casting instruction right? In the compiled asm there's an addition `scvtf` if using a non-constant cast.

GCC doesn't apply the transform if `y` is not a constant: https://godbolt.org/z/qTd6Tnh3c
If the `powi` is only used by the `log`, `sitofp + fmul` is definitely cheaper than a call to `pow`.

Comment (tanmaysachan) [reply llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:2507]: Got it! Will update PR  with the changes.

Comment (tanmaysachan) [reply llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp]: Added sitofp conversion for non-consts.

**Comment by tanmaysachan** [comment]
Ping @dtcxzyw, any other suggestion?

File: llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:2510]:
```suggestion
        Y = B.CreateCast(Instruction::SIToFP, Y, Ty, "cast");
```
`IRBuilder` will handle the constant folding.
https://github.com/llvm/llvm-project/blob/21fe8b635cfb04de486a3c4ca22eb9f5a5ed78ad/llvm/include/llvm/IR/IRBuilder.h#L2138-L2139

**Review by dtcxzyw — APPROVED** [review]
LGTM. Please wait for additional approval from other folks.

File: llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp
Comment (tanmaysachan) [reply llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp]: Fixed.

File: llvm/test/Transforms/InstCombine/log-pow.ll
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/log-pow.ll:22]: Requiring fast for this transform seems like overkill, but that's a pre-existing issue that could be relaxed

File: llvm/test/Transforms/InstCombine/log-pow.ll
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/log-pow.ll:36]: Can you also test with the f32 case, and some cases with missing flags?

Comment (tanmaysachan) [reply llvm/test/Transforms/InstCombine/log-pow.ll:36]: Added tests for f32 and non-fast.

File: llvm/test/Transforms/InstCombine/log-pow.ll
Comment (tanmaysachan) [reply llvm/test/Transforms/InstCombine/log-pow.ll:22]: Is there any existing issue about shifting some transformers to non-fast? I could have a look at it.

File: llvm/test/Transforms/InstCombine/log-pow.ll
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/log-pow.ll:45]: what about llvm.powi + llvm.log, not the lib call?

Comment (tanmaysachan) [reply llvm/test/Transforms/InstCombine/log-pow.ll:57]: Added a test with explicit calls to both.

**Comment by tanmaysachan** [comment]
ping @nikic @jcranmer-intel 
Could you review this?

File: llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp
Comment (nikic) [inline llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:2507]:
```suggestion
      Y = B.CreateSIToFP(Y, Ty, "cast");
```

**Review by nikic — APPROVED** [review]

File: llvm/test/Transforms/InstCombine/log-pow.ll
Comment (andykaylor) [inline llvm/test/Transforms/InstCombine/log-pow.ll:17]: This is a bit of a problem, though not one introduced by this PR. The call to log() may set errno in some circumstances, but the call to llvm.log isn't required to do so. If you compile with clang, it will call llvm.log in place of log() when math errno support isn't required. You can get clang to set fast on a call to the log library function using "-ffast-math -fmath-errno" but I'm not sure what that means since the fast-math flags allow us to replace the function.

I'm being pedantic here, and I don't mean for this to block the PR, but this is the primary difference between the library function and the intrinsic, so it seems worth thinking about.

Comment (tanmaysachan) [reply llvm/test/Transforms/InstCombine/log-pow.ll:17]: @andykaylor Since this is my 2nd PR to llvm, I'm slightly out of my depth here. Are you proposing I drop all the tests with a call to log() directly?

Comment (arsenm) [reply llvm/test/Transforms/InstCombine/log-pow.ll:17]: The math library function and intrinsic situation is a bit of a mess, and we kind of try to solve the same issue in 2-3 different places. Clang both swaps out the intrinsic for the libcall based on -fno-math-errno, and codegen also tries to specially recognize libcall+readnone callsites as ignore-errno. 

Ideally here we could perform the transform from libcall->libcall and intrinsic->intrinsic, but should avoid mutating between the two

File: llvm/test/Transforms/InstCombine/log-pow.ll
Comment (arsenm) [reply llvm/test/Transforms/InstCombine/log-pow.ll:22]: This feels like reassoc?

File: llvm/test/Transforms/InstCombine/log-pow.ll
Comment (andykaylor) [reply llvm/test/Transforms/InstCombine/log-pow.ll:17]: > @andykaylor Since this is my 2nd PR to llvm, I'm slightly out of my depth here. Are you proposing I drop all the tests with a call to log() directly?

No. The tests should stay. I was just pointing out that the transformation shouldn't be converting the libcall to an intrinsic, as @arsenm described. That's not something that needs to be changed in the PR though. It's an existing problem.

File: llvm/test/Transforms/InstCombine/log-pow.ll
Comment (andykaylor) [reply llvm/test/Transforms/InstCombine/log-pow.ll:22]: There was a discussion a few years ago about some place where we were checking reassoc when what we really wanted to do was distribution. I don't remember what the outcome was. As I said, I'd really like to have a broadly defined "algebraic-identity" flag, but we're out of bits (as currently implemented).

I think we agree that "-ffast-math" (and -funsafe-math-optimizations) should definitely allow transformations like this, but once you start turning off individual flags (which is important for users who are trying to locally restrict fast-math to solve accuracy problems), it's hard to say which combination should be required from the current set.

With the current "isFast()" check, even disabling contraction blocks the transformation, which is surely not what we want.

Comment (jcranmer-intel) [reply llvm/test/Transforms/InstCombine/log-pow.ll:22]: The current trend is using `reassoc` for these sorts of transformations, but I'm not entirely happy with that logic. I'm putting together a RFC on tweaking the fast math flags and bringing some more clarity to these flags.

File: llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp
Comment (arsenm) [inline llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:2498]: could add fix for libcall/intrinsic mismatch problem

**Review by arsenm — APPROVED** [review]

**Pull request merged**
